KEP-2170: Add Kubeflow Trainer Pipeline Framework Design (#2439)
kubeflow/trainer: docs/proposals/2170-kubeflow-training-v2/README.md

_Will be added after initial implementation for PyTorch._

> ## Pipeline Framework

Author: Yeah, sure. In that case, where do we want to put this?
I am guessing if we should create `Kubeflow Trainer > User Guide > PlatformDeveloper Guide`.

> ## Pipeline Framework

Reviewer: I think, we should add them under Operator Guides: https://www.kubeflow.org/docs/components/trainer/operator-guides/

> ## Pipeline Framework

Author: Ah, I didn't find that. That sounds great.

> ## Pipeline Framework

Author: Done: https://github.com/kubeflow/trainer/issues/2458

We introduce the framework as internal mechanism so that we can easily expand mechanism
for combination of Runtimes and TrainJob.

The framework is called as Kubeflow Trainer Pipeline Framework, and it has 4 phases as you can see the following
overview.

![Overview](./TrainerPipelineFrameworkOverview.drawio.svg)

As described in the following, each phase is basically executed step by step although `Startup Phase` is executed only once
during starting trainer-controller-manager:

> - `Startup Phase`: Initialize internal components at once when the trainer-controller-manager starts.

Author: SGTM

> - `Startup Phase`: Initialize internal components at once when the trainer-controller-manager starts.

Author: Done.

- `PreExecution Phase`: This phase is executed as a part of admission validating webhooks triggered by TrainJob is created and updated.
- `Build Phase`: This phase is executed to build child Kubernetes resources and deploy those to the cluster.
- `PostExecution Phase`: This phase is executed after the `Build Phase`.

As you can see in the diagram, each phase has 2 types of APIs, `Internal API` and `Extension Point`.

> The Extension Point is exposed and could be added operations within the scope of the Pipeline Framework Plugins Interfaces as plugins
> and those plugins are performed in any order.

Reviewer 2: ```suggestion
The Extension Point is exposed, allowing operations to be added as plugins within the scope of the Pipeline Framework Plugins Interfaces. These plugins can be executed in any order.
```
Rephrase this sentence so it would be more readable:)

Author: Actually, `These plugins can be executed in any order.` is incorrect since we can not handle the order.
So, I will refine this with `These plugins are executed in any order.`

Author: Done.

On the other hand, the Internal APIs are not exposed and could not add any operations as opposed to the Extension Point.

![Kubeflow TrainerPipelineFramework](./TrainerPipelineFramework.drawio.svg)

> - `Startup Phase`:

Reviewer: I am wondering if we should also say that cluster operators can:
1. Register new plugins into existing TrainingRuntime and ClusterTrainingRuntime
2. Register new runtimes into runtime framework (e.g. SlurmRuntime).
Or you think, it is out-of-scope since we don't support registration of new runtimes yet ?

Author: > Register new runtimes into runtime framework (e.g. SlurmRuntime).
Or you think, it is out-of-scope since we don't support registration of new runtimes yet ?

Yeah, the current framework does not support inserting arbitrary plugins and runtimes (for sure, technically, we could do it). However, I think we can add another InternalAPI `Initialize Trainer Framework Pipelines` to diagram to clarify when pipelines are initialized. WDYT?

Reviewer: Yeah, I think that would be nice.

Author: Done.

    and triggering TrainJob reconciliations.
- `PreExecution Phase`:
  - Extension Point:
    - `CustomValidation`: This registers validators for validating any kind of resources to Admission Validating Webhook Servers
    when TrainJob is created and updated.
- `Build Phase`:
  - Internal API:
    - `ComponentDeployer`: This deploys built components (resources) to the cluster which is performed as a part of reconciler.
  - Extension Point:
    - `EnforcePodGroupPolicy`: This configures PodGroup specific parameters (e.x, specified in TrainingRuntime `.spec.podGroupPolicy`)
    to any kind of resources like PodSpec.
    - `EnforceMLPolicy`: This configure MachineLearning framework specific parameters (e.x, specified in TrainingRuntime `.spec.mlPolicy`)
    to any kind of resources like PodSpec.
    - `ComponentBuilder`: This builds Kubernetes resources leveraging `RuntimeInfo` and `TrainJob`.
    `RuntimeInfo` is abstracted objects extracted from runtimes like TrainingRuntime and ClusterTrainingRuntime.
- `PostExecution Phase`:
  - Internal API:
    - `SupendedCondition`: Check if TrainJob is suspended state, and then add `Suspended` condition to TrainJob.

>     - `CreatedConditon`: Check if TrainJob is created state, and then add `Created` condition to TrainJob.

Reviewer 3: Maybe more a general comment rather than something strictly in the scope of this PR, I find the name of this condition a bit confusing as the TrainJob has to be "created", even before that condition can be applied.

So either it could be stated that this condition is about whether the children components of the TrainJob have been created, or possibly rename the condition to something like `InitializedCondition`.

Reviewer 3: > What about trying to change "Created" condition type name in a follow-up PR?

That sounds good to me 👍🏼.

Reviewer: Sounds good, let's discuss it separately.
@astefanutti So do you suggest that we split Created condition for TrainJob between two ?
https://github.com/kubeflow/trainer/blob/8a6091907a2df6ab1d55b9713ead631f72ece56e/docs/proposals/2170-kubeflow-training-v2/README.md#state-transition

Reviewer 3: @andreyvelich I was more simply thinking about renaming it to something like `Initialized` or `ComponentsCreated` more than splitting it. As it stands it may be interpreted as the TrainJob is created, which is not what it is.

Reviewer: @astefanutti Please can you create an issue so we can continue discussion on how we can improve the TrainJob conditions ?
We just need to make sure that we don't confuse users given that we also have TrainJob Initializers.

Author: @astefanutti @andreyvelich opened: https://github.com/kubeflow/trainer/issues/2459
Let's track this enhancement in a dedicated issue.
